helm/manifest_json.go

```diff
@@ -0,0 +1,91 @@
+package helm
+
+import (
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/yaml"
+)
+
+var yamlDocumentSeparator = "\n---"
+
+type resourceMeta struct {
+	metav1.TypeMeta
+	Metadata metav1.ObjectMeta
+}
+
+func convertYAMLManifestToJSON(manifest string) (string, error) {
+	m := map[string]json.RawMessage{}
+	resources := strings.Split(manifest, yamlDocumentSeparator)
+	for _, resource := range resources {
+		jsonbytes, err := yaml.YAMLToJSON([]byte(resource))
+		if err != nil {
+			return "", fmt.Errorf("could not convert manifest to JSON: %v", err)
+		}
+
+		resourceMeta := resourceMeta{}
+		err = yaml.Unmarshal([]byte(resource), &resourceMeta)
+		if err != nil {
+			return "", err
+		}
+
+		gvk := resourceMeta.GetObjectKind().GroupVersionKind()
+		key := fmt.Sprintf("%s/%s", strings.ToLower(gvk.GroupKind().String()),
+			resourceMeta.Metadata.Name)
+
+		if namespace := resourceMeta.Metadata.Namespace; namespace != "" {
+			key = fmt.Sprintf("%s/%s", namespace, key)
+		}
+
+		if gvk.Kind == "Secret" {
+			secret := corev1.Secret{}
+			err = yaml.Unmarshal([]byte(resource), &secret)
+			if err != nil {
+				return "", err
+			}
+
+			for k, v := range secret.Data {
+				// compute a checksum for the secret field so it appears in the diff
+				sha := sha512.New()
+				sha.Write(v)
+				sum := sha.Sum(nil)
+				secret.Data[k] = []byte(fmt.Sprintf("(sensitive value %x)", sum[:8]))
+			}
+
+			jsonbytes, err = json.Marshal(secret)
+			if err != nil {
+				return "", err
+			}
+		}
+
+		m[key] = jsonbytes
```

Comment:
Just some context here – we have to munge the parsed manifest into a slightly different structure to capture the namespace/kind/name as a key at the top of the structure, otherwise this information gets hidden by Terraform's concise diff feature and makes it slightly confusing to read when there is many resources in the manifest.